### PR TITLE
Support for DAGRUN and DAGRUN_RO 

### LINKED ISSUE #3: Support for AI.DAGRUN_RO
https://oss.redislabs.com/redisai/commands/#aidagrun_ro

### PR DESCRIPTION
**Pull request by filipecosta90**
This PR adds support for DAGRUN and DAGRUN_RO commands ( including the newly added SCRIPTRUN within the DAG ).
- fixes #3 
- fixes #4
- Includes positive and negative testing for dagrun and dagrun_ro. Regarding the positive testing, we've also added a larger test with ResNet-50 example with no pre or post-processing.

### CODE REVIEW

**Comment by filipecosta90** [comment]
@dengliming I was not able to ping on the reviewers. I would be very happy if chimed in and share your thoughts about it =) ( p.s: still adding docs and an example of it )

File: src/dag.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+export class Dag implements DagCommandInterface {
+  private _commands: any[][];
+  private _tensorgetflag: boolean[];
+
+  constructor() {
+    this._commands = [];
+    this._tensorgetflag = [];
+  }
+
+  public tensorset(keName: string, t: Tensor) {
```
Comment (dengliming) [inline src/dag.ts:30]: Can we change to chained calls here? like: `public tensorset(keName: string, t: Tensor) : Dag ` and `ends with return this;`, Not sure if this is supported : )

File: src/client.ts [hunk 2]
```diff
@@ -151,6 +116,38 @@ export class Client {
       });
   }
 
+  /**
+   *
+   * @param loadKeys
+   * @param persistKeys
+   * @param dag
+   */
+  public dagrun(loadKeys: string[] | null, persistKeys: string[] | null, dag: Dag): Promise<any> {
```
Comment (dengliming) [inline src/client.ts:126]: This seems that we can do add a field `readonly `on `Dag`, Then we can merge `dagrun `and `dagrun_ro `into one method. WDYT?  :)

Comment (filipecosta90) [reply src/client.ts:126]: I believe both have pros and cons. I'm fine with both :+1: . lets see what @lantiga or @hhsecond think about it :)

File: src/dag.ts (hunk 1 above)
Comment (filipecosta90) [reply src/dag.ts]: Yes! great idea. Added the changes and an test case with it:
```
dag
      .tensorset('tensorA', tensorA)
      .tensorset('tensorB', tensorB)
      .modelrun('mymodel-dag', ['tensorA', 'tensorB'], ['tensorC'])
      .tensorget('tensorC');
```
Thank you @dengliming =)

File: src/dag.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import { Model } from './model';
+import { Script } from './script';
+import { Tensor } from './tensor';
+
+export interface DagCommandInterface {
+  tensorset(keName: string, t: Tensor): DagCommandInterface;
```
Comment (dengliming) [inline src/dag.ts:6]: Mybe we should change `keName` to `keyName` and keep consistent with other places？

Comment (filipecosta90) [reply src/dag.ts]: agree. Fixed in the latest commit

File: src/client.ts (hunk 2 above)
Comment (filipecosta90) [reply src/client.ts:126]: @gkorland can you share your thoughts on both options?

File: src/tensor.ts [hunk 4]
```diff
@@ -84,4 +99,8 @@ export class Tensor {
     }
     return new Tensor(dt, shape, values);
   }
+
+  static tensorGetFlatArgs(keyName: string): any[] {
```
Comment (gkorland) [inline src/tensor.ts:103]: why any[] and not string[]?

File: src/script.ts [hunk 5]
```diff
@@ -87,4 +87,28 @@ export class Script {
...
+      args.push('TAG');
+      args.push(this.tag);
+    }
+    args.push('SOURCE');
+    args.push(this.script);
+    return args;
+  }
+
+  static scriptRunFlatArgs(scriptName: string, functionName: string, inputs: string[], outputs: string[]) {
+    const args: any[] = [scriptName, functionName, 'INPUTS'];
```
Comment (gkorland) [inline src/script.ts:103]: string[]

File: src/script.ts [hunk 6]
```diff
@@ -87,4 +87,28 @@ export class Script {
...
+    inputs.forEach((value) => args.push(value));
+    args.push('OUTPUTS');
+    outputs.forEach((value) => args.push(value));
+    return args;
+  }
+
+  static scriptGetFlatArgs(scriptName: string) {
+    const args: any[] = [scriptName, 'META', 'SOURCE'];
```
Comment (gkorland) [inline src/script.ts:111]: string[]

File: src/model.ts [hunk 7]
```diff
@@ -132,4 +132,35 @@ export class Model {
...
+  static modelRunFlatArgs(modelName: string, inputs: string[], outputs: string[]): any[] {
+    const args: any[] = [modelName, 'INPUTS'];
+    inputs.forEach((value) => args.push(value));
+    args.push('OUTPUTS');
+    outputs.forEach((value) => args.push(value));
+    return args;
+  }
+
+  modelSetFlatArgs(keyName: string) {
+    const args: any[] = [keyName, this.backend.toString(), this.device];
```
Comment (gkorland) [inline src/model.ts:149]: string[]

File: src/model.ts [hunk 8]
```diff
@@ -132,4 +132,35 @@ export class Model {
     }
     return model;
   }
+
+  static modelGetFlatArgs(keyName: string): any[] {
+    return [keyName, 'META', 'BLOB'];
+  }
+
+  static modelRunFlatArgs(modelName: string, inputs: string[], outputs: string[]): any[] {
+    const args: any[] = [modelName, 'INPUTS'];
```
Comment (gkorland) [inline src/model.ts:141]: string[]

File: src/tensor.ts (hunk 4 above)
Comment (filipecosta90) [reply src/tensor.ts:103]: fixed in the latest commit

File: src/script.ts (hunk 5 above)
Comment (filipecosta90) [reply src/script.ts:103]: solved in latest commit

File: src/script.ts (hunk 6 above)
Comment (filipecosta90) [reply src/script.ts:111]: solved in latest commit

File: src/model.ts (hunk 7 above)
Comment (filipecosta90) [reply src/model.ts:149]: for this one we need to keep the []any, given that argument of type 'Buffer' is not assignable to parameter of type 'string'

File: src/model.ts (hunk 8 above)
Comment (filipecosta90) [reply src/model.ts:141]: solved in the latest commit

File: src/client.ts (hunk 2 above)
Comment (filipecosta90) [reply src/client.ts:126]: hi there @dengliming I believe we should keep the dagrun and dagrun_ro to have a direct association with the RedisAI commands. WDYT?

Comment (dengliming) [reply src/client.ts:126]: agree. : )

Comment (filipecosta90) [reply src/client.ts:126]: :+1: thanks for the fast reply =)

**Pull request merged**